Add doc from oneofixx form.

njohner opened the pull request:
Add menu action to create a document from a oneoffixx template:
* Form to choose a template, with filtering per template groups
* Document created in shadow state, with template_id in the annotations
* Registry entry to turn feature on/off

Some comments:
* Do we need the button to decide whether to open the doc un officeconnector after creation? It probably doesn't make sense to create the document without openning it in officeconnector
* So maybe we should only allow the feature when OfficeConnector is also turned on?


![screen shot 2018-04-13 at 09 11 21](https://user-images.githubusercontent.com/7374243/38722330-ec8436a0-3efd-11e8-9b20-049cdf7a0253.png)


The request to the OneOffix API is not part of the PR and will be implemented in a separate PR, the API request results, are currently mocked.

resolves #4134

phgross commented on opengever/document/oneoffixx/command.py:
```diff
@@ -0,0 +1,27 @@
+        self.filename = template.filename
+        super(CreateDocumentFromOneOffixxTemplateCommand, self).__init__(
+            context, title)
+        # if we need an empty file...
```
Not sure if we need an empty File, @Rotonen ?

phgross commented on opengever/document/oneoffixx/form.py:
```diff
@@ -0,0 +1,182 @@
+                     oneofixx_template(template),
+                     str(template.get("template-id")),
+                     template.get("title")))
+    if template_group is not None and template_group[0] != '--NOVALUE--':
```
I would add a comment here like `filter templates when template_group has been selected`

phgross commented on opengever/document/oneoffixx/form.py:
```diff
@@ -0,0 +1,182 @@
+    return MutableObjectVocabulary(terms)
+
+
+class oneofixx_template(object):
```
- Class Names always in " 🐫 - case"
 - Typo in of**f**ixx

`OneOffixTemplate`

phgross commented on opengever/document/oneoffixx/form.py:
```diff
@@ -0,0 +1,182 @@
+         'created': date(2018, 2, 3),
+         'group': 'Briefe'}
+    ]
+    return templates
```
Should return a list of OneOffixxTemplate objects ...

phgross commented on opengever/document/oneoffixx/form.py:
```diff
@@ -0,0 +1,182 @@
+        required=True)
+
+    form.widget(edit_after_creation=SingleCheckBoxFieldWidget)
+    edit_after_creation = schema.Bool(
```
`edit_after_creation` field is not necessary

phgross commented on opengever/document/oneoffixx/form.py:
```diff
@@ -0,0 +1,182 @@
+    def finish_document_creation(self, data):
+        new_doc = self.create_document(data)
+
+        if data.get('edit_after_creation'):
```
✂️

phgross commented on opengever/document/oneoffixx/form.py:
```diff
@@ -0,0 +1,182 @@
+        manager = self.context.restrictedTraverse('checkout_documents')
+        manager.checkout(new_doc)
+
+        new_doc.setup_external_edit_redirect(self.request)
```
Instead of the `checkout` action it should use the new action `oneoffixx` (see #4136) as soon it's implemented.

I would extend the `setup_external_edit_redirect` method with an keyword argument `action`.

phgross commented on opengever/core/profiles/default/registry.xml:
```diff
@@ -82,6 +82,9 @@
   <!-- OFFICEATWORK -->
   <records interface="opengever.officeatwork.interfaces.IOfficeatworkSettings" />
 
+  <!-- ONEOFFIXX -->
```
Please also extend the rest api config endpoint (see https://github.com/4teamwork/opengever.core/blob/master/opengever/api/config.py)

phgross commented on opengever/oneoffixx/tests/test_feature.py:
```diff
@@ -0,0 +1,38 @@
+from plone import api
+
+
+class TestIsOneoffixxFeatureEnabled(FunctionalTestCase):
```
Use IntegrationTests. And register a feature `oneoffixx` (see https://github.com/4teamwork/opengever.core/blob/master/opengever/testing/integration_test_case.py#L37-L53)

phgross commented on opengever/oneoffixx/tests/test_oneoffixx.py:
```diff
@@ -0,0 +1,59 @@
+        browser.fill({'Edit after creation': False})
+        browser.find('Save').click()
+
+        shadow_doc = self.dossier.restrictedTraverse('document-27', None)
```
The browser should redirect to the document, so i would use `browser.context` instead of `document-27` otherwise the test will fail as soon an additional object is added in the fixture.

phgross commented on opengever/oneoffixx/tests/test_oneoffixx.py:
```diff
@@ -0,0 +1,59 @@
+from opengever.testing import IntegrationTestCase
+
+
+class TestCreateDocFromOneoffixxTemplate(IntegrationTestCase):
```
There is a test missing which checks that the template-ID of the selected template is stored in the annotations of the document.

phgross commented on opengever/oneoffixx/locales/de/LC_MESSAGES/opengever.oneoffixx.po:
```diff
@@ -0,0 +1,50 @@
+#. Default: "Template group"
+#: ./opengever/oneoffixx/form.py
+msgid "label_template_group"
+msgstr "Vorlage Gruppe"
```
Vorlagengruppe

njohner commented:
All done. I put the corrections in a separate commit, so it's a bit easier to follow.  Main changes are:
* Add a test to check that the template ID is in the annotations of the document
* Only allow the `oneoffixx` action when both the oneoffixx and the officeconnector features are enabled
* Add `action` kwarg to `setup_external_edit_redirect`. When `action=="oneoffixx"` we do nothing for now

njohner commented:
I think this is ready for now. 
* Obviously connection to the api is missing.
* @Rotonen I looked at the OC api tests, but not sure what we'd gain from that for now? We first need to implement the OC stuff (payload and such)?

Rotonen commented:
@njohner correct. They only test OC interactionability and represent the blind assumptions of OC.

njohner commented on setup.py:
```diff
@@ -43,10 +43,9 @@
-          open("README.rst").read() + "\n" +
-          open(os.path.join("docs", "HISTORY.txt")).read()
-          ),
+      long_description=("%s\n%s" % (open("README.rst").read(),
```
Pycodestyle was shooting here because of the line break after a "+" sign. If I added a "\" he'd shoot because it's not needed as the statement is in brackets...

Rotonen replied:
We recently version bumped, which changed some of the subtleties. Rather amend the ignoreset in the buildouts.

https://github.com/4teamwork/ftw-buildouts/blob/3a7dc172a2e908697318ba372c37c10e4519b1bc/pycodestyle.cfg

njohner replied:
From the conversation referenced in your commit in `ftw-buildouts`, it seems that line-break before the operator is better style. So instead of ignoring the warning for line-break after binary operator I just shifted the linebreak here.

njohner commented on opengever/oneoffixx/configure.zcml:
```diff
@@ -0,0 +1,23 @@
+      for="*"
+      name="document_with_oneoffixx_template"
+      class=".form.SelectOneOffixxTemplateDocumentView"
+      permission="zope2.View"
```
Would probably make more sense

njohner commented on opengever/oneoffixx/form.py:
```diff
@@ -0,0 +1,201 @@
+def get_templates_from_api():
+    templates = [
+                 {'templates': [
+                                {u'languages': [2055],
```
The real API gave numbers. This is data from the real API

njohner commented on opengever/oneoffixx/configure.zcml:
```diff
@@ -0,0 +1,23 @@
+  <i18n:registerTranslations directory="locales" />
+
+  <browser:page
+      for="*"
```
Restricted to `IDossierMarker`

njohner commented on opengever/oneoffixx/form.py:
```diff
@@ -0,0 +1,201 @@
+    return templates
+
+
+def get_oneoffixx_templates():
```
I renamed the functions and used a generator comprehension, it's more readable and efficient that way

njohner commented:
Ready for re-review.
* Pinned the version of `requests-oauthlib`
* Moved browser views in a separate directory
* Rebased and moved API config
* Changed permissions and registration for the view to generate document from oneoffixx template.

Note that simple generation from templates uses the `zope2.View` and `for="*"` (https://github.com/4teamwork/opengever.core/blob/master/opengever/dossier/templatefolder/configure.zcml#L6-L11)

njohner commented:
I rewrote the history and removed the `requests-oauthlib`which we don't yet used.
